### PR TITLE
Make running-time calculation causal

### PR DESCRIPTION
**Pull request by axrolld**
OSRD's running-time calculation algorithm is not causal : envelope parts are computed in an order that's convenient but not chronological. Braking curves are computed, then accelerating curves, then constant speed parts.

This make impossible implementing a battery train, because in that case we need to know how much energy has been spent to know which traction force is available.

This enabler is making OSRD's running-time calculation algorithm causal, computing the different envelope parts in the right order.

### CODE REVIEW

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java [hunk 1]
```diff
@@ -367,12 +344,6 @@ public boolean findSpeed(double speed, CmpOperator operator) {
                 getPartBound(part, operator),
                 operator, speed);
 
-        // look for the next envelope part which contains a speed which matches the
-        // search requirement. The current envelope part may contain such a point,
-        // but it may be **before the cursor**.
-        if (!findPart(partPredicate))
-            return false;
-
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java]: Is there a reason for this change? It sounds equivalent so I don't really care either way, but I may be missing something.

The comment in the following `while` becomes outdated with this change.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java [hunk 2]
```diff
@@ -48,50 +47,51 @@ public static Envelope addAccelerationCurves(
...
-            double inertia = rollingStock.getInertia();
-            double worstRamp = Math.asin((maxTractionForce - rollingResistance) / inertia / 9.81) * 1000;
-            var envelopePart = cursor.getPart();
-            while (!cursor.hasReachedEnd() && cursor.getPart() == envelopePart) {
-                double highRampPosition = path.findHighGradePosition(
-                        cursor.getPosition(), envelopePart.getEndPos(), rollingStock.getLength(), worstRamp);
-                cursor.findPosition(highRampPosition);
-                if (cursor.getPosition() == envelopePart.getEndPos())
-                    break;
+        while (!cursor.hasReachedEnd()){
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: No space between `)` and `{`? I'm surprised this doesn't cause a checkstyle error

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java [hunk 3]
```diff
@@ -34,8 +33,9 @@ public boolean initCheck(double direction, double position, double speed) {
         cursor = new EnvelopeCursor(envelope, direction < 0, partIndex, stepIndex, position);
         if (type == CEILING)
             return envelopeSpeed >= speed;
-        else
+        if (type == FLOOR)
             return envelopeSpeed <= speed;
+        return true;
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java]: At this point a `switch` would probably be better, the third enum value would becomes explicit. It's not clear when reading this code that the last `return` is only reached if `type == MAINTAIN_SPEED`

Always returning `true` with a certain type of envelope seems weird. Is there a reason?

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopePartConstraintType.java [hunk 4]
```diff
@@ -3,4 +3,5 @@
 public enum EnvelopePartConstraintType {
     FLOOR,
     CEILING,
+    MAINTAIN_SPEED,
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopePartConstraintType.java]: What does `MAINTAIN_SPEED` mean in this context? Is it a constraint that "fails" if the envelope differs in either direction? If that's the case, I don't think the name is very clear about it.

A comment would help clarify this either way.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/SpeedConstraint.java [hunk 5]
```diff
@@ -20,8 +19,11 @@ public SpeedConstraint(double speedConstraint,
     public boolean initCheck(double direction, double position, double speed) {
         if (type == CEILING)
             return speed <= speedConstraint;
-        else
+        if (type == FLOOR)
             return speed >= speedConstraint;
+        if (type == MAINTAIN_SPEED)
+            return speed == speedConstraint;
+        return true;   // default return
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/SpeedConstraint.java]: With a switch, there's no need for a default return.

(This "please use a switch" comment also applies to the rest of the `if` chains with enums)

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java [hunk 6]
```diff
@@ -10,25 +9,25 @@
...
-        if (part.stepCount() != 1)
-            return false;
-        return part.getBeginSpeed() == part.getEndSpeed();
+        return part.getMinSpeed() == part.getMaxSpeed();
     }
 
-    /** Generate acceleration curves overlay everywhere the max speed envelope increase with a discontinuity */
-    public static Envelope addAccelerationCurves(
+    /** Generate acceleration curves overlay everywhere the max speed envelope increase with a discontinuity
+     * and compute the constant speed parts to check if the train can physically maintain its speed or not */
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]:
```suggestion
    /** Generate acceleration curves overlay everywhere the max speed envelope increase with a discontinuity
     * and compute the constant speed parts to check whether the train can physically maintain its speed */
```

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java [hunk 7]
```diff
@@ -48,50 +47,51 @@ public static Envelope addAccelerationCurves(
...
+        while (!cursor.hasReachedEnd()){
+            if (cursor.checkPart(MaxEffortEnvelope::maxEffortPlateau)){
+                var partBuilder = new EnvelopePartBuilder();
+                partBuilder.setAttr(EnvelopeProfile.CONSTANT_SPEED);
+                var startSpeed = cursor.getStepBeginSpeed();
+                var startPosition = cursor.getPosition();
+                var overlayBuilder = new ConstrainedEnvelopePartBuilder(
+                        partBuilder,
+                        new SpeedConstraint(startSpeed, MAINTAIN_SPEED),
+                        new SpeedConstraint(0, FLOOR),
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: This constraint is probably redundant with MAINTAIN_SPEED

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java [hunk 8]
```diff
@@ -48,50 +47,51 @@ public static Envelope addAccelerationCurves(
...
+                        new PositionConstraint(cursor.getPart().getBeginPos(), cursor.getPart().getEndPos())
+                );
+                EnvelopeMaintain.maintain(context, startPosition, startSpeed, overlayBuilder, 1);
+                // Check if the speed can be maintained
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: This comment should probably be moved after the following `if`, it doesn't seem related (if I follow what is going in)

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java [hunk 9]
```diff
@@ -48,50 +47,51 @@ public static Envelope addAccelerationCurves(
...
+                    startPosition = cursor.getPosition();
+                    startSpeed = maxSpeedProfile.interpolateSpeedLeftDir(startPosition, 1);
+                    EnvelopeAcceleration.accelerate(context, startPosition, startSpeed, overlayBuilder, 1);
+                    cursor.findPosition(overlayBuilder.getLastPos());
 
+                    // Check that the high grade position can't be maintained
+                    if (partBuilder.stepCount() > 1)
+                        builder.addPart(partBuilder.build());
+                    else
+                        cursor.findPosition(cursor.getPosition()  + 1);
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: A comment would help here as well, it seems like we "skip" a small part.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java [hunk 10]
```diff
@@ -101,14 +101,9 @@ public static Envelope addMaintainSpeedCurves(
...
-
-                // Check that the high grade position can't be maintained
-                if (partBuilder.stepCount() > 1)
-                    builder.addPart(partBuilder.build());
-                else
-                    cursor.findPosition(cursor.getPosition()  + 1);
+                builder.addPart(partBuilder.build());
             }
-            cursor.nextPart();
+            else cursor.nextPart();
```
Comment (eckter) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]:
```suggestion
            else
                cursor.nextPart();
```

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java [hunk 11]
```diff
@@ -78,6 +77,31 @@ void testSpeedCeilingConstraint() {
...
+    @Test
+    void testSpeedMaintainConstraint() {
+        var partBuilder = new EnvelopePartBuilder();
+        var constraint = new SpeedConstraint(2, MAINTAIN_SPEED);
+        var builder = new ConstrainedEnvelopePartBuilder(
+                partBuilder,
+                constraint
+        );
+        assertTrue(constraint.initCheck(1, 2, 2));
+        assertTrue(builder.initEnvelopePart(2, 2, 1));
```
Comment (eckter) [inline core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java]: Not a comment on the PR itself, you may ignore this

We have two methods that do mostly the same thing, one is called using `(direction, position, speed)`, the other `(position, speed, direction)`. This feels very wrong.

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java [hunk 12]
```diff
@@ -78,6 +77,31 @@ void testSpeedCeilingConstraint() {
...
+        assertTrue(builder.addStep(2.5, 2));
+        assertFalse(builder.addStep(3, 1));
+        assertEquals(0, builder.lastIntersection);
```
Comment (eckter) [inline core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java]: We test that it stops if we go below the limit, but we don't test what happens if it goes above the limit

**Review by eckter — COMMENTED** [review]
Nice job

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java (hunk 1 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java]: It was failing a test, and i could not find a reason for this specific block of code. Removing it was fixing the test so i just did and no other test failed. I might have fixed what this specific block was about with this PR.

I'll fix the comment

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java (hunk 2 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: Fixed !

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java (hunk 3 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java]: The `true` default value is when no constraint type has been sent, which does not happen. It's just that calling this method without a constraint type doesn't really make sense, because without constraint the check is always `true`.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopePartConstraintType.java (hunk 4 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopePartConstraintType.java]: It means the constraint is to maintain the given speed. It indeed fails if the speed differs from the given speed.

I'm open to rename it, i'm just not very inspired right now.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java (hunk 6 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: Done

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java (hunk 7 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: You're right, i'll remove it

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java (hunk 8 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: I changed the comment, hope it's clearer now

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java (hunk 9 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: I'll add some comment to clarify it ! Let me know if that helps

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java (hunk 10 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxEffortEnvelope.java]: Fixed

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java (hunk 11 above)
Comment (axrolld) [reply core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java]: I'll just remove this as it was to make codeCov happy, but yeah i'll re-arrange the order of the args

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java (hunk 12 above)
Comment (axrolld) [reply core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/ConstraintBuilderTest.java]: I'm adding a line

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java (hunk 1 above)
Comment (eckter) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java]: There's likely a bug hidden somewhere, this block shouldn't do anything other than an optimization. It skips forward until it finds a part that can match the criteria.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java (hunk 3 above)
Comment (eckter) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java]: I don't think there can be a case where no constraint type has been set. It seems like it's a fix to the "not every path returns something" error, which hides the problem when a new enum value is added. This can also be fixed with a switch.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java (hunk 1 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java]: I can add it back, but it seemed very weird to me

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java (hunk 3 above)
Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/constraints/EnvelopeConstraint.java]: I left a switch without default value

**Review by bgiuliana — DISMISSED** [review]

**Review by eckter — DISMISSED** [review]

**Review by eckter — DISMISSED** [review]

**Review by eckter — APPROVED** [review]

**Review by bgiuliana — APPROVED** [review]
